### ISSUE TITLE
Ability to use url parameters for optionally filtering down

### ISSUE THREAD

**Issue by vesper8** [wdelfuego/nova-calendar#72]
I'm using 2.x and I have a calendar that shows aggregated data through `nonNovaEvents`

it all works great, but I would like to be able to optionally open this calendar with a url parameter in order to only display events related to id passed in the url

So.. default calendar: `/nova/calendar/lessons-summary`

But sometimes via clicking on a Nova resource I want to open `/nova/calendar/lessons-summary?bookingId=1`

And then in my calendar I have

```
        $query = Booking::with([
            'bookingLessons.bookingLessonDetails',
        ])
        ->whereBetween('starts_at', [
            $this->startOfCalendar(),
            $this->endOfCalendar()
        ])
        ->orderBy('starts_at', 'asc');

        $bookings = $query->get();
```

Now I would have liked to simply add a 

```
        if(request()->has('bookingId')) {
            $query->where('id', request('bookingId'));
        }
```

But unfortunately while I'm inside the calendar I can't seem to gain access to the url parameters via the request object. I tried both `request()->all()` and `request()->route()` and neither seem to expose my url parameter.

I'm not entirely sure why that is, I'm guessing it's because the calendar provider is not so much a controller but powers the api.

I was wondering if you have any insight on how I may be able to achieve the desired result? Many thanks!

**Comment by wdelfuego** [comment]
Thanks for the clear issue!

You can't access those request parameters, because they are part of the url you are requesting the Nova app with, while the request for calendar data is a separate AJAX request to the API, that doesn't automatically include request parameters from the main page load.

For something like this to work, we'd need a way to be able to send custom context parameters to the calendar API call, and to get them from the url of the calendar view page within Nova. This entails changes in both the backend routes and the front-end JS logic which constructs the calendar data API url.

There's no easy fix, but It'd be a nice feature to be able to parametrize the calendar view like that, so I might look into offering this functionality as a general feature in the future.

For now;
- You can try to get this working by forking and customizing the package
- You could try to solve this with event filtering; the problem is then reduced to automatically activating an event filter based on the calendar's url in Nova. (this still requires forking)
- A nasty work-around which doesn't require forking could be to generate independent calendar views automatically for each individual booking in `config/nova-calendar.php`. That could allow you to construct a bunch of calendar view urls using the booking-id as part of the url, and link to them directly. You then simply don't add all of those calendar views to the Menu, but they should still work when directly linked to. You can use the same CalendarDataProvider for all of them and deduce the booking id from the requested url since it is then part of the API request url.

**Comment by vesper8** [comment]
@wdelfuego Thank you for the quick and detailed response : )

I think it would be a nice feature to add in a future release.

I've figured out another way to do it which didn't involve forking.. and I think it works quite well.. I only have one little problem with it.. let me show you how I did it.

First I installed this package https://github.com/sietse85/nova-button

Then I added a new button on my resource:
```
Button::make('Inspect in Calendar')->action(InspectInCalendar::class),
```

This is my action:
```
<?php

namespace App\Nova\Actions;

use Cache;
use Laravel\Nova\Actions\Action;
use Illuminate\Support\Collection;
use Laravel\Nova\Fields\ActionFields;

class InspectInCalendar extends Action
{
    public function handle(ActionFields $fields, Collection $models)
    {
        Cache::set('calendar-booking-id', $models[0]->id, 60);

        return Action::redirect('/nova/calendar/lessons-summary');
    }
}
```

Then inside my CalendarProvider:

```
        if(Cache::has('calendar-booking-id')) {
            $query->where('id', Cache::get('calendar-booking-id'));

            Cache::forget('calendar-booking-id');
        }
```


And voila! It works great! Easy peasy.

The only problem is that if my booking is in December, I cannot guarantee that it will open the Calendar on the month of December. It seems that the Calendar will default to showing the current month, and then it will somehow remember where it left off, so if you open the calendar, and click the previous button a bunch of times and go back to March.. then next time you open the Calendar again it will still be in March. This is a convenient feature for most cases. But in this use-case above, which is almost already perfect.. I need to be able to somehow tell the Calendar to open on a specific month. Is there any way to do this currently? Either through a URL parameter or maybe using the Cache helper? Ideally something that doesn't involve javascript or forking.

Many thanks!!

**Comment by vesper8** [comment]
It seems like `AbstractCalendarDataProvider` has the methods I would need to set the calendar start and end that I want based on Cached values.. but I've tried using 

```
    public function __construct()
    {
        $this->startOfCalendar(\Illuminate\Support\Carbon::parse('March 2023'));
        $this->endOfCalendar(\Illuminate\Support\Carbon::parse('April 2023'));

        parent::__construct();
    }
```

And have tried putting these methods inside my `nonNovaEvents()` and I can't get it to work, I think the values get overwritten by the request object no matter what.

**Comment by vesper8** [comment]
Seems like I can just override the `startOfCalendar` and `endOfCalendar` methods themselves by placing them in my own CalendarProvider, and from there I can return whatever range I want which does sort of result in the behaviour I want, however it doesn't affect the main label at the top and it probably breaks other things... I'm sure there's a better way... I'll wait for your reply : ) Thanks!

**Comment by wdelfuego** [comment]
Ooh, that's a nice work-around to get this working! Much better than what I thought of. Thanks for sharing 👍.

If I understand correctly, you want the calendar to open on the month of the current booking, if a 'currently selected id' is found in the cache. You're basically 'abusing' a single calendar view to behave as different calendars to the end user, so the current behavior you describe is desired and intended - we want a single calendar view to maintain its selected year and month between page loads.

The front-end is responsible for maintaining the selected year and month (from localStorage), but the back-end can choose to override the requested year/month data and return calendar data for a different year and month.

However, the `CalendarDataProvider` is not the place to do so; you should instead look into `Wdelfuego\NovaCalendar\Http\Controllers\CalendarController::getCalendarData` and the `Wdelfuego\NovaCalendar\View\Month::setYearAndMonth` methods.

I would probably try to tackle it like this:
- Implement your own subclass of Wdelfuego\NovaCalendar\Http\Controllers\CalendarController
- Make your app use your custom controller instead of the default one using some Laravel dependency injection voodoo
- Copy the entire implementation of the `getCalendarData(string $view = 'month')` method, but just before returning `$view->calendarData(...)` at the end of the method, call `$view->setYearAndMonth(...)` and pass it the year and month for the currently selected booking.

However, that effectively disables manual navigation through the years and months, since the booking's year and month will always be selected. Not sure if that's what you want. Also, this might not work well with your `Cache::forget` call since you might need to rely on the same cached value in multiple places. I'm also not sure how this work-around would work out if you are also showing other, "normal" calendars in your application. This might break those. If this calendar is the only instance of this package in your app I think a setup like this could work fine.

Come to think of it, I think it might be a lot easier if I released a small update that allows your `CalenderDataProvider` to customize the view just before generating the result data. For example, I could update the CalendarController to call something like `CalendarDataProvider::customizeView($view)` at the right point and then you could simply implement this logic within that method in your CalendarDataProvider. But this could still effectively disable manual navigation because the back-end has no way of knowing if the user is just opening a calendar for the first time or manually trying to navigate between years and months. You could probably work around this by setting another Cache flag when the button is clicked and then forgetting that once it's used, so the year and month are only manually overridden once when a new booking is selected.

I'd be happy to release an update today or tomorrow to facilitate something like this. 

Let me know what your thoughts are :)!

**Comment by vesper8** [comment]
Hey thanks for your great response again! Yea.. I use multiple calendars the "normal" way and only in this one instance I want to allow my client to be able to click a single booking from its' Nova resource page and it should load the calendar on that booking's month and display only that one booking. Because of the Cache::forget, if they use the calendar navigation to go previous/next month, or if they simply reload the page, then the calendar should revert to its normal usage and display all events as though the cache had never been set.

Yea your initial workaround seems quite complex and would not only break the normal usage but would also probably break with future updates of your package too. Your alternative solution to add the means of setting the desired month/year of the calendar inside the custom calendar data provider is much cleaner and flexible.

I guess the way I would then handle it is instead of using a single cache variable, I'd use an additional one or two to pass the desired month/year. I'd forget the cache values after they've been used. The end result should be that after clicking the button in the Nova resource page it would load the desired month/year and only display that booking, but any reload or navigation would just revert to the normal usage with everything displayed.

Hope that makes sense. Much appreciated!

**Comment by wdelfuego** [comment]
OK, update incoming

**Comment by wdelfuego** [comment]
v2.1.3 has just been released.

You can now implement `CalendarDataProvider::customizeView(View $view) : View` to manipulate the view before output under certain conditions (your Cache flag system).

In that method, you can call the new method `View::forceShowDate(Carbon $date)`. Send it your booking's date value to make sure the calendar shows the proper date range for that booking. You can't send a year or month directly because the View system is prepared for other views like a week view, too. A `forceShowDate` method is the generic approach to what would otherwise be called `forceYearAndMonth` or something.

It's been a bit of a hotfix but I'm convinced it doesn't break anything; in case it does, or if it doesn't work as intended, please shoot me a message ASAP 🙏.

Good luck!

**Comment by vesper8** [comment]
@wdelfuego Awesome!! Thank you so much!

I think there's a bug though

I think this line inside CalendarController `$view = $dataProvider->customizeView($this);` needs to be `$view = $dataProvider->customizeView($view);`

If I make this change then all works well, otherwise I get this error `LessonsSummaryDataProvider::customizeView(): Argument #1 ($view) must be of type Wdelfuego\NovaCalendar\View\AbstractView, Wdelfuego\NovaCalendar\Http\Controllers\CalendarController given`

**Comment by wdelfuego** [comment]
Correct, thanks, fixed in 2.1.4

**Comment by vesper8** [comment]
Works great!! Thanks so much! Going to close this issue now. Cheers!
